src/common/utxobased/engine/makeUtxoEngineState.ts

```diff
@@ -447,18 +430,37 @@ export function makeUtxoEngineState(
       }
     },
     async processUtxos(utxos: IUTXO[]) {
-      const utxoMap: Map<string, Set<IUTXO>> = new Map()
+      // Map of scriptPubkey to IUTXOs
+      const utxoMap: Map<string, IUTXO[]> = new Map()
+      const utxoIds: Set<string> = new Set()
+
+      // Map updated utxos
       for (const utxo of utxos) {
-        const utxoSet: Set<IUTXO> = utxoMap.get(utxo.scriptPubkey) ?? new Set()
+        const utxoSet = utxoMap.get(utxo.scriptPubkey) ?? []
         if (!utxoMap.has(utxo.scriptPubkey))
           utxoMap.set(utxo.scriptPubkey, utxoSet)
-        utxoSet.add(utxo)
+        // Add utxo to set for scriptPubkey
+        utxoSet.push(utxo)
+        // Add UTXO ID set
+        if (utxo.spent) utxoIds.add(utxo.id)
       }
+
+      // Process UTXO sets for each scriptPubkey in map
       for (const [scriptPubkey, utxos] of utxoMap.entries()) {
+        // Get saved utxo set
+        const savedUtxos = await processor.fetchUtxos({ scriptPubkey })
+        // Filter UTXOs to de-duplicate (and undefined utxo as a type assert)
+        const filteredUtxos = savedUtxos.filter(
+          utxo => utxo != null && !utxoIds.has(utxo?.id)
+        ) as IUTXO[]
+
+        // Add updated utxos to utxo set
+        const combinedUtxos = [...filteredUtxos, ...utxos]
```

Comment:
Can a 'non-spent' utxo in appear in both the 'filteredUtxos' and the 'utxos' objects?
And if so, wouldn't that create duplications on 'combinedUtxos'?

Reply:
Yes, and I changed this last night to filter all IDs in the `utxos` set. This way there is no dupes.